Update release process docs

amyblais opened the pull request:
- Cleaned up release docs to remove any outdated checklists since we are using Playbooks to run checklists.
 - Updated release docs to reflect that we are now in a bi-weekly release cadence for Cloud releases (instead of a weekly cadence).

amyblais committed:
Update release-overview.md

amyblais committed:
Update release-overview.md

amyblais committed:
Update cloud-release.md

amyblais committed:
Update dot-release.md

amyblais committed:
Update playbooks-release.md

amyblais committed:
Update security-release.md

amyblais committed:
Update mobile-release.md

amyblais committed:
Update feature-release.md

amyblais committed:
Update desktop-release.md

amyblais committed:
Update desktop-release.md

amyblais committed:
Update feature-release.md

amyblais committed:
Update mobile-release.md

lindalumitchell commented on operations/research-and-development/product/release-process/release-overview.md:
```diff
@@ -1,6 +1,6 @@
 # Release Overview
 
-Mattermost ships with a new version on the 16th of each month in [binary form](https://docs.mattermost.com/upgrade/upgrading-mattermost-server.html#upgrading-to-the-latest-version) for self-managed customers. Mattermost also ships releases with new features and bug fixes [to Mattermost Cloud](https://docs.mattermost.com/about/product.html#mattermost-cloud) weekly.
+Mattermost ships with a new version on the 16th of each month in [binary form](https://docs.mattermost.com/upgrade/upgrading-mattermost-server.html#upgrading-to-the-latest-version) for self-managed customers. Mattermost also ships releases with new features and bug fixes to [Mattermost Cloud](https://docs.mattermost.com/about/cloud-subscriptions.html) weekly.
```
```suggestion
Mattermost ships with a new version on the 16th of each month in [binary form](https://docs.mattermost.com/upgrade/upgrading-mattermost-server.html#upgrading-to-the-latest-version) for self-managed customers. Mattermost also ships releases with new features and bug fixes to [Mattermost Cloud](https://docs.mattermost.com/about/cloud-subscriptions.html) biweekly.
```

lindalumitchell commented on operations/research-and-development/product/release-process/cloud-release.md:
```diff
@@ -6,33 +6,18 @@ Please refer to [the Cloud Release Playbook](https://community.mattermost.com/pl
+ - T-14 to T-3: 
+    - Run Rainforest release test run groups, Cypress automated tests and product high level release smoke tests (i.e. Boards, Calls, Channels, Integration Frameworks, Playbooks and Suite Users). Close or label tickets for the release. Test failure reviews and bug fixing.
+ - T-3 (Monday)
+    - Each QA team signs off as appropriate.
```
```suggestion
    - Each QA signs off as appropriate, using the [QA Approval playbook](https://community.mattermost.com/playbooks/playbooks/qeay15ou6frsmnu6tpmup9mxbc) run for that release.
```

lindalumitchell commented on operations/research-and-development/product/release-process/desktop-release.md:
```diff
@@ -10,168 +10,12 @@ Please refer to [the desktop app release playbook](https://community.mattermost.
+ - Cut release branch based off `master` and cut RC-1 (around T-14).
+ - Code Freeze at T-5.
+ - Cut Final build at T-2.
+    - QA approval should be ready by T-2.
```
```suggestion
    - [QA approval](https://community.mattermost.com/playbooks/playbooks/h798dt39mpbymb8z5uoiuf4hdo/outline) should be ready by T-2.
```

lindalumitchell commented on operations/research-and-development/product/release-process/mobile-release.md:
```diff
@@ -8,135 +8,12 @@ Please refer to [the Mobile Release Playbook](https://community.mattermost.com/p
+ - Cut release branch based off `main` and cut beta builds (around T-14).
+ - Code Freeze at T-5.
+ - Cut Final build at T-2.
+    - QA approval should be ready by T-2.
```
```suggestion
    - [QA approval](https://community.mattermost.com/playbooks/playbooks/8dx9zrs16bbg8eg1hdhw35zdua) should be ready by T-2.
```

lindalumitchell commented on operations/research-and-development/product/release-process/feature-release.md:
```diff
@@ -1,232 +1,20 @@
+ - Code Freeze at T-5.
+    - Prepackaged plugins should be ready on or before this date.
+ - Cut Final build at T-2.
+    - QA approval should be ready by T-2.
```
```suggestion
    - [QA approval](https://community.mattermost.com/playbooks/playbooks/8dx9zrs16bbg8eg1hdhw35zdua) should be ready by T-2.
```

amyblais committed:
Apply suggestions from code review

Co-authored-by: lindalumitchell <linda@mattermost.com>

amyblais committed:
Update operations/research-and-development/product/release-process/feature-release.md

Co-authored-by: lindalumitchell <linda@mattermost.com>

lindalumitchell reviewed (dismissed):
Looks great, thanks Amy!

saturninoabril commented on operations/research-and-development/product/release-process/cloud-release.md:
```diff
@@ -6,33 +6,18 @@ Please refer to [the Cloud Release Playbook](https://community.mattermost.com/pl
+    - Run Rainforest release test run groups, Cypress automated tests and product high level release smoke tests (i.e. Boards, Calls, Channels, Integration Frameworks, Playbooks and Suite Users). Close or label tickets for the release. Test failure reviews and bug fixing.
+ - T-3 (Monday)
+    - Each QA signs off as appropriate, using the [QA Approval playbook](https://community.mattermost.com/playbooks/playbooks/qeay15ou6frsmnu6tpmup9mxbc) run for that release.
+    - QA approval should be given by 5pm Eastern on Monday so that the release rollout can be started early Tuesday morning. 
```
Not by 5pm but it's the latest, right?
```suggestion
    - QA approval should be given latest by 5pm Eastern on Monday so that the release rollout can be started early Tuesday morning. 
```

saturninoabril commented on operations/research-and-development/product/release-process/desktop-release.md:
```diff
@@ -10,168 +10,12 @@ Please refer to [the desktop app release playbook](https://community.mattermost.
-    - Post and review [Mattermost Security Updates](https://mattermost.com/security-updates/) for the Desktop App
-    - Update Security Issues spreadsheet with issue number from posted update (e.g. v3.2.0.1)
-    - Confirm the Security Researchers list on the [Responsible Disclosure Policy](https://mattermost.com/security-vulnerability-report/) is up to date
+Note: T-minus counts are measured in "working days" (weekdays other than major holidays concurrent in US and Canada) prior to release day.
```
Not related to this line but for the line above regarding the quarterly ship.

> Desktop App releases follow a fixed schedule of 4 releases per year. Releases ship quarterly by the 16th of March, June, September, and December of each year.

I wonder if 16th of Dec is a good timing to do the release given during this period (before and after) most people are taking holiday breaks and the support maybe limited. Thoughts @m1lt0n @amyblais?

amyblais replied:
I think releasing in December is fine, but I'll let @m1lt0n and @devinbinnie comment if they have additional thoughts.

m1lt0n replied:
Initially that schedule was created to help spread the QA reviews of the internal QA people. Since the situation has change, we're open to a different schedule (we'd prefer to stick to the fixed cadence, but if different months e.g. Mid-April, July, October and January make more sense now, we're open to it)

@devinbinnie Any preference or adjustments we want to make here?

saturninoabril replied:
Thanks @m1lt0n! Sounds good to me. @lindalumitchell thoughts?

lindalumitchell replied:
Thanks for asking! Given that people being out in mid- and late December might crunch testing time for a January release (possibly even more than a December release), I'd vote for either leaving it as-is, or shifting it in the other direction, e.g. Feb, May, Aug, Nov. But I'm fine either way.

saturninoabril replied:
Seems like all are fine as is, and so I'm good with it as well.

saturninoabril commented on operations/research-and-development/product/release-process/playbooks-release.md:
```diff
@@ -2,7 +2,7 @@
 This document outlines the Playbooks release process for the Mattermost core team. Our release process draws from many best practice Agile, Scrum, and Software Development Lifecycle methodologies.
 
-Every two weeks, the team runs through a playbook to cut a new release of the plugin, smoke test it, and add it to the marketplace. That gets pushed to cloud immediately \(via split.io\), and any self-managed customers who desire the upgrade may do so from the plugin marketplace.
+Every month, the team runs through a playbook to cut a new release of the plugin, smoke test it, and add it to the marketplace. That gets pushed to cloud immediately \(via split.io\), and any self-managed customers who desire the upgrade may do so from the plugin marketplace.
```
Playbooks has recently moved to `mattermost-server`, built-in as product. I wonder if plugin release is still accurate or has to be updated. Same case about the [Boards release doc](https://github.com/mattermost/mattermost-handbook/blob/amyblais-patch-2/operations/research-and-development/product/release-process/boards-release.md). Maybe something for @sbishel to confirm?

amyblais replied:
Monorepo didn't make it to v7.10 release, so I'll let Scott to confirm how this works now.

sbishel replied:
Other than security patches, neither boards nor playbooks will cut plugins any more.

saturninoabril commented on operations/research-and-development/product/release-process/release-overview.md:
```diff
@@ -28,11 +28,11 @@ Mattermost numbers stable releases in the following format: **\[Version Number\]
 ## Multi-product Release Principles
 
-The goal is not to have the same release cadence for all products, but there may be similar schedules. E.g. Playbooks and Boards plugins are pre-packed once a month with the self-managed releases, however, admins will have a config option to turn off products. As the product suite gets bigger, we need a dedicated Release Manager for each product. We can duplicate/reuse a similar build process for ...
+The goal is not to have the same release cadence for all products, but there may be similar schedules. E.g. Playbooks and Boards plugins are pre-packed once a month with the self-managed releases, however, admins will have a config option to turn off products. We can duplicate/reuse a similar build process for new plugins in the suite.
```
Same here in pre-packaging Playbooks and Boards if still accurate given they are to be shipped as built-in products.

amyblais replied:
Monorepo didn't make it to v7.10 release, so I'll let @sbishel to confirm how this works now.

sbishel replied:
Other than security patches, neither boards nor playbooks will cut plugins any more.

saturninoabril commented on operations/research-and-development/product/release-process/release-overview.md:
```diff
@@ -71,28 +71,34 @@ The goal is not to have the same release cadence for all products, but there may
+    - Run Rainforest release test run groups, Cypress automated tests and product high level release smoke tests (i.e. Boards, Calls, Channels, Integration Frameworks, Playbooks and Suite Users). Close or label tickets for the release. Test failure reviews and bug fixing.
+ - T-3 (Monday)
+    - Each QA team signs off as appropriate.
+    - QA approval should be given by 5pm Eastern on Monday so that the release rollout can be started early Tuesday morning. 
```
```suggestion
    - QA approval should be given latest by 5pm Eastern on Monday so that the release rollout can be started early Tuesday morning. 
```

saturninoabril reviewed (commented):
Thanks @amyblais for the quick updates! Overall looks good to me, except for few clarifications.

enahum commented on operations/research-and-development/product/release-process/mobile-release.md:
```diff
@@ -8,135 +8,12 @@ Please refer to [the Mobile Release Playbook](https://community.mattermost.com/p
+
+**Schedule for Mobile App releases**:
+ - Feature Complete deadline is approximately 1 month prior to the release day.
+ - Cut release branch based off `main` and cut beta builds (around T-14).
```
I would add here to cut an RC based on the release branch and start testing. here is the reason

1. the weekly beta builds are always based of main branch and can be used for progressing testing until we branch for release, once branched, beta could have code that is not going to be part of the release.
2. Give more time to QA to run their test suite (automated and manual)
3. Identify problems with the release early

Thoughts @saturninoabril

amyblais replied:
I agree with this suggestion, I'll also add @larkox to share any additional thoughts.

saturninoabril replied:
Thanks @enahum and I totally agree. I honestly thought that we normally do progressive testing during release period for mobile app. I didn't realize that "the weekly beta builds are always based of main branch" is the case.

amyblais replied:
Also adding @lindalumitchell for any thoughts.

lindalumitchell replied:
Thanks, I also agree, and apparently had confusion as well. :)

enahum commented on operations/research-and-development/product/release-process/mobile-release.md:
```diff
@@ -8,135 +8,12 @@ Please refer to [the Mobile Release Playbook](https://community.mattermost.com/p
+**Schedule for Mobile App releases**:
+ - Feature Complete deadline is approximately 1 month prior to the release day.
+ - Cut release branch based off `main` and cut beta builds (around T-14).
+ - Code Freeze at T-5.
```
here I would add cutting another RC based of the release branch

enahum reviewed (dismissed):
I'm going to approve as my suggestions will require some changes with the QA team that should be discussed first.

amyblais committed:
Update operations/research-and-development/product/release-process/cloud-release.md

Co-authored-by: Saturnino Abril <saturnino.abril@gmail.com>

amyblais committed:
Update operations/research-and-development/product/release-process/release-overview.md

Co-authored-by: Saturnino Abril <saturnino.abril@gmail.com>

amyblais committed:
Update release-overview.md

amyblais committed:
Update SUMMARY.md

larkox commented on operations/research-and-development/product/release-process/mobile-release.md:
```diff
@@ -8,135 +8,12 @@ Please refer to [the Mobile Release Playbook](https://community.mattermost.com/p
+
+**Schedule for Mobile App releases**:
+ - Feature Complete deadline is approximately 1 month prior to the release day.
+ - Cut release branch based off `main` and cut beta builds (around T-14).
```
Just for my knowledge. What is the difference between being in feature complete, and having the release branch cut? My main concern is that if no features get into the code after T-28, but the release branch is not cut until T-14, we would be "blocking" the merge of features those 14 days... every month. Shouldn't feature complete be at the same time as we cut the release branch? (I may be misunderstanding or mixing concepts, so please bear with me)

amyblais replied:
Makes sense, I've edited it so that feature complete is at T-14.

larkox commented on operations/research-and-development/product/release-process/mobile-release.md:
```diff
@@ -8,135 +8,12 @@ Please refer to [the Mobile Release Playbook](https://community.mattermost.com/p
+
+**Schedule for Mobile App releases**:
+ - Feature Complete deadline is approximately 1 month prior to the release day.
+ - Cut release branch based off `main` and cut beta builds (around T-14).
```
Once the release branch is created, how often do we want to create builds (release candidates of sort) for QA? We can keep creating betas from main, but as stated before, once the release branch is created, the beta (out of main) may contain code that is not meant to release. So on those 14 days since the release branch is created, at least 1 beta will be released out of sync with the release.

In the meanwhile, some fixes may be cherry picked to the release branch.

We could create a new build after every cherry pick, or a more forgiving cadence (once a day if there has been any merge, or even once a week). 0/5 on the final details, but I feel we should at least discuss it.

amyblais replied:
Adding @lindalumitchell & @saturninoabril for any thoughts.

lindalumitchell replied:
Thanks, I think at most one new build per day for testing (if there have been changes / cherry picks of course) would be often enough.

saturninoabril replied:
> Once the release branch is created, how often do we want to create builds (release candidates of sort) for QA?

+1 to what Linda said.

@larkox How do we know if the build is from main or release branch during the release testing period or in general? How about alpha for main and beta build for release branch?

larkox replied:
I will leave this to @enahum , since changing the name of the app to support this, may not be possible (or practical). We could potentially add a new parameter to the build to use on the build info (build name?) along the build number, build version, etc... But not sure that is the best idea neither.

enahum replied:
Oh boy, this is getting out of hand and will create a lot of confusion, it already started with community members.

Cutting builds have an impact, specially on the stores so having a cadence of daily builds is not something that I will encourage.

Having a 3rd app is also not an option, 2 should be more than enough as it allows us to have the latest and gratest as well as what is actually being shipped.

The idea is to cover our based but without over complicating things.

As I mentioned in the QA channel, the idea is to coordinate between the teams to make minimize the pain, not to add more :) 

That said, we can build as many as we want like PR's that do not increase the build numbers nor are submitted to the stores (including testflight)

saturninoabril replied:
I understand and agree to limit into 2 apps (beta and release), and definitely agree about team coordination especially during release testing. Also noted about the daily build. It may not likely to happen during the early phase of release testing but may be requested as we get closer to the deadline in a case where a critical bug fix is necessary to ship.

Currently, when beta builds are posted at `Release: Mobile Apps`, there's an info about `Version` and `Build Number` ([example](https://community.mattermost.com/core/pl/m7uhfq1g1tyabrj958686cgjxc)). How about adding extra fields such as `Branch` and `Commit (short sha)` so that we, in QA, can easily identify the base branch where it's built?

amyblais replied:
I've made edits to [the Mobile release process section](https://github.com/mattermost/mattermost-handbook/pull/1192/files#diff-2589aadc650609878898e858a1cd2ce69dd31bba3c001f94be7182d8f797a031) based on the discussions here.

larkox reviewed (dismissed):
With the changes proposed by Elias, LGTM. Nevertheless, there are two minor discussion point open, but I don't want to block this PR based on those.

amyblais committed:
Update mobile-release.md

lindy65 committed:
Update release-overview.md

Changed "biweekly" to "bi-weekly"

lindy65 committed:
Update release-overview.md

Changed "biweekly" to "bi-weekly" in line 3 (missed in previous commit)

lindy65 committed:
Update release-overview.md

Changed "... targeted to.." to "... targeted at..." in line 111

lindy65 committed:
Update release-overview.md

Changed "A backup of the `cloud` branches ARE saved" to "A backup of the `cloud` branches IS saved" in line 137

lindy65 reviewed (dismissed):
Thanks @amyblais - made a couple of spelling/grammar updates. Otherwise looks good to me 👍

lindalumitchell reviewed (dismissed):
Thanks, LGTM!

amyblais committed:
Update mobile-release.md

amyblais committed:
Update mobile-release.md

saturninoabril reviewed (approved):
Thanks @amyblais, LGTM!
